IdentityCore/src/cache/metadata/MSIDMetadataCache.m

```diff
@@ -0,0 +1,166 @@
+                                     MSIDErrorInvalidInternalParameter,
+                                     @"cacheItem and key could not be nil.",
+                                     nil, nil, nil, nil, nil);
+        }
+        return NO;
+    }
+    
+    __block NSError *localError;
+    __block BOOL saveSuccess = NO;
+    
+    if ([item isEqual:_memoryCache[key]]) return YES;
+    
+    dispatch_barrier_sync(_synchronizationQueue, ^{
+        saveSuccess = [_dataSource saveAccountMetadata:item key:key serializer:_jsonSerializer context:context error:&localError];
+        if (saveSuccess)
+        {
+            _memoryCache[key] = item;
+        }
+    });
+    
+    if (error) *error = localError;
+    return saveSuccess;
+}
+
+- (MSIDAccountMetadataCacheItem *)accountMetadataWithKey:(MSIDCacheKey *)key
+                                                 context:(id<MSIDRequestContext>)context
+                                                   error:(NSError **)error
+{
+    if (!key)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain, MSIDErrorInternal, @"Key is not valid.", nil, nil, nil, context.correlationId, nil);
+        }
+        MSID_LOG_ERROR(context, @"Set keychain item with invalid key.");
+        return nil;
+    }
+
+    __block MSIDAccountMetadataCacheItem *item;
+    __block NSError *localError;
+
+    dispatch_sync(_synchronizationQueue, ^{
+        item = _memoryCache[key];
+        if (!item)
+        {
+            item = [_dataSource accountMetadataWithKey:key serializer:_jsonSerializer context:context error:&localError];
+        }
+    });
+    
+    if (error) *error = localError;
+    return item;
+}
+
+- (BOOL)removeAccountMetadataForKey:(MSIDCacheKey *)key
+                            context:(id<MSIDRequestContext>)context
+                              error:(NSError **)error
+{
+    if (!key)
+    {
+        if (error)
+        {
+            *error = MSIDCreateError(MSIDErrorDomain,
+                                     MSIDErrorInvalidInternalParameter,
+                                     @"cacheItem and key could not be nil.",
+                                     nil, nil, nil, nil, nil);
+        }
+        return NO;
+    }
+    
+    __block BOOL success = NO;
+    __block NSError *localError;
+    
+    dispatch_sync(_synchronizationQueue, ^{
+        [_memoryCache removeObjectForKey:key];
+        
+        success = [_dataSource removeAccountMetadataForKey:key context:context error:&localError];
+    });
+    
+    if (error) *error = localError;
+    return success;
+}
+
+- (BOOL)clearWithContext:(id<MSIDRequestContext>)context
+                   error:(NSError **)error
+{
+    __block BOOL success = NO;
+    __block NSError *localError;
+    
+     dispatch_sync(_synchronizationQueue, ^{
+         [_memoryCache removeAllObjects];
+         
+         success = [_dataSource clearWithContext:context error:&localError];
```

Comment:
this will clear everything including tokens, right?

Reply:
depending on the dataSource, but for our context, by default, MSIDKeychainTokenCache is commonly used for accounts and metadata -
short answer, YES for us in iOS. 
NO/MAYBE/YES for macOS or is initiated with different datasource.